// @sap/ux-odata-client/dist/secure-store/keytarStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class KeytarStore {
    constructor(log, keytar) {
        this.log = log;
        this.keytar = keytar;
    }
    async save(service, key, value) {
        try {
            const serialized = JSON.stringify(value);
            await this.keytar.setPassword(service, key, serialized);
            return true;
        }
        catch (e) {
            this.log.error(`Error saving to secure store. Service: [${service}], key: [${key}]`);
            this.log.error(e);
            return false;
        }
    }
    async retrieve(service, key) {
        try {
            const serializedValue = await this.keytar.getPassword(service, key);
            return JSON.parse(serializedValue);
        }
        catch (e) {
            this.log.error(`Error retrieving from secure store. Service: [${service}], key: [${key}]`);
            this.log.error(e);
            return null;
        }
    }
    async delete(service, key) {
        try {
            return await this.keytar.deletePassword(service, key);
        }
        catch (e) {
            this.log.error(`Error deleting from secure store. Service: [${service}], key: [${key}]`);
            this.log.error(e);
            return false;
        }
    }
    async getAll(service) {
        try {
            return (await this.keytar.findCredentials(service)).map((x) => JSON.parse(x.password));
        }
        catch (e) {
            this.log.error(`Error getting values for service: [${service}]`);
            this.log.error(e);
            return [];
        }
    }
}
exports.KeytarStore = KeytarStore;
