// @sap/ux-odata-client/src/client/httpClient.ts
// eslint-disable-next-line import/no-duplicates
import axios, { AxiosInstance } from 'axios';
import { connect, Connection } from '../connection';
import { Logger } from '../utils';
import { SystemConfig, Credentials } from '../config';

interface HttpClientCreateOptions {
    system: SystemConfig;
    credentials: Credentials;
    log: Logger;
    existingConnection?: Connection;
}

export async function newHttpClient({
    system,
    credentials,
    log,
    existingConnection
}: HttpClientCreateOptions): Promise<{ connection: Connection; httpClient: AxiosInstance }> {
    const connection = existingConnection || (await connect(system, credentials, log));
    const headers = {
        Cookie: connection.cookies.toString()
    };

    if (connection.xsrfToken) {
        headers['x-csrf-token'] = connection.xsrfToken;
    }

    const httpClient = axios.create({
        baseURL: new URL(system.service, system.url).toString(),
        headers,
        withCredentials: true
    });
    return { connection, httpClient };
}
