// @sap/ux-specification/dist/src/sync/v2/export/pages/ListReport.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../../common/decoration/decorators");
const decorators_2 = require("../../../common/decoration/decorators");
const decoration_1 = require("../../../common/decoration");
const types_1 = require("../../../common/types");
let ListReport = class ListReport {
};
__decorate([
    decorators_2.exportRule({
        flex: {
            controlId: decoration_1.buildControlIdForPage,
            controlType: () => types_1.ControlType.DynamicPage,
            exportFunction: decoration_1.exportToFlexChange
        }
    })
], ListReport.prototype, "fitContent", void 0);
__decorate([
    decorators_1.childId('listReportFilter')
], ListReport.prototype, "filterBar", void 0);
__decorate([
    decorators_1.childId('listReport')
], ListReport.prototype, "table", void 0);
ListReport = __decorate([
    decorators_1.pageLayoutInformation({
        id: 'sap.suite.ui.generic.template.ListReport.view.ListReport',
        pageId: 'sap.suite.ui.generic.template.ListReport'
    })
], ListReport);
exports.ListReport = ListReport;
