// @sap/ux-odata-client/dist/system/systemStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("./index");
const secure_store_1 = require("../secure-store");
const config_1 = require("../config");
class DefaultImplementation {
    constructor(log) {
        this.secureStore = secure_store_1.getSecureStore(log);
    }
    async getAll() {
        const systems = {};
        const serializedSystems = await this.secureStore.getAll(config_1.ServiceName.SystemInfo);
        serializedSystems.forEach((s) => {
            const sys = index_1.SapSystem.fromSerialized(s);
            systems[sys.name] = sys;
        });
        return systems;
    }
    async get(id) {
        const serializedSystem = await this.secureStore.retrieve(config_1.ServiceName.SystemInfo, id);
        if (serializedSystem) {
            return index_1.SapSystem.fromSerialized(serializedSystem);
        }
        else {
            return undefined;
        }
    }
    async save(system) {
        return this.secureStore.save(config_1.ServiceName.SystemInfo, system.getId(), system.serialize());
    }
    async delete(system) {
        return this.secureStore.delete(config_1.ServiceName.SystemInfo, system.getId());
    }
}
exports.getSystemStore = (log = console) => {
    return new DefaultImplementation(log);
};
