// @sap/ux-specification/dist/src/sync/v2/export/controls/FilterBar.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../../common/decoration/decorators");
const decoration_1 = require("../../../common/decoration");
const application_1 = require("../../application");
const types_1 = require("../../../common/types");
class FilterBar {
}
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    }),
    decorators_1.validity({
        since: '1.38.0'
    })
], FilterBar.prototype, "smartVariantManagement", void 0);
__decorate([
    decorators_1.exportRule({
        flex: {
            controlId: decoration_1.buildControlIdFromParent,
            controlType: () => types_1.ControlType.SmartFilterBar,
            exportFunction: decoration_1.exportToFlexChange
        }
    })
], FilterBar.prototype, "liveMode", void 0);
exports.FilterBar = FilterBar;
