// @sap/ux-specification/dist/src/sync/v2/export/controls/Table.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../../common/decoration/decorators");
const decoration_1 = require("../../../common/decoration");
const application_1 = require("../../application");
const types_1 = require("../../../common/types");
const getSmartTableControlId = (baseId, idBreadcrumbs, breadcrumbs, controlType) => {
    if (controlType() === 'sap.m.Table') {
        return baseId + 'responsiveTable';
    }
    else if (controlType() === 'sap.ui.comp.smarttable.SmartTable' && baseId.includes('AnalyticalListPage')) {
        return baseId + 'table';
    }
    return decoration_1.buildControlIdFromParent(baseId, idBreadcrumbs);
};
class DataloadSettings {
}
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getDataLoadSettingPathV2(pageKeys)
        }
    }),
    decorators_1.validity({
        since: '1.75.2'
    })
], DataloadSettings.prototype, "loadDataOnAppLaunch", void 0);
exports.DataloadSettings = DataloadSettings;
class CommonTableSettings {
    constructor() {
        this.getDeprecated = () => {
            return [
                {
                    tableType: {
                        since: '1.50.0',
                        deprecated: {
                            since: '1.70.0',
                            text: 'Use tableSettings.type instead.'
                        }
                    }
                },
                {
                    multiSelect: {
                        since: '1.38.0',
                        deprecated: {
                            since: '1.70.0',
                            text: 'Use tableSettings.multiSelect instead'
                        }
                    }
                },
                {
                    enableAutoBinding: {
                        since: '1.50.0',
                        deprecated: {
                            since: '1.75.2',
                            text: 'Use tableSettings.loadDataOnAppLaunch instead.'
                        }
                    }
                }
            ];
        };
    }
}
__decorate([
    decorators_1.exportRule({
        flex: {
            controlId: decoration_1.buildControlIdFromParent,
            controlType: () => types_1.ControlType.SmartTable,
            exportFunction: decoration_1.exportToFlexChange
        }
    })
], CommonTableSettings.prototype, "showTablePersonalisation", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: () => 'sap.ui.generic.app/settings'
        }
    }),
    decorators_1.validity({
        since: '1.63.0'
    })
], CommonTableSettings.prototype, "showDraftToggle", void 0);
__decorate([
    decorators_1.exportRule({
        flex: {
            controlId: getSmartTableControlId,
            controlType: () => types_1.ControlType.SmartTable,
            exportFunction: decoration_1.exportToFlexChange
        }
    })
], CommonTableSettings.prototype, "useExportToExcel", void 0);
__decorate([
    decorators_1.exportRule({
        flex: {
            controlId: getSmartTableControlId,
            controlType: () => types_1.ControlType.SmartTable,
            exportFunction: decoration_1.exportToFlexChange
        }
    })
], CommonTableSettings.prototype, "fitContainer", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    }),
    decorators_1.validity({
        since: '1.48.0'
    })
], CommonTableSettings.prototype, "quickVariantSelection", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys),
            export: (manifestSection, configObject) => {
                if (manifestSection['quickVariantSelectionX']) {
                    for (const variant in configObject['variants']) {
                        const currentQuickVariantX = configObject['variants'][variant];
                        currentQuickVariantX['tableSettings'] = {};
                        const tablesettings = ['type', 'multiSelect', 'selectAll', 'selectionLimit'];
                        for (const key in currentQuickVariantX) {
                            if (tablesettings.indexOf(key) !== -1) {
                                currentQuickVariantX['tableSettings'][key] = currentQuickVariantX[key];
                                delete currentQuickVariantX[key];
                            }
                        }
                    }
                }
                manifestSection['quickVariantSelectionX'] = configObject;
            }
        }
    }),
    decorators_1.validity({
        since: '1.48.0'
    })
], CommonTableSettings.prototype, "quickVariantSelectionX", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getTableSettingPathV2(pageKeys)
        }
    })
], CommonTableSettings.prototype, "multiSelect", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getTableSettingPathV2(pageKeys)
        }
    })
], CommonTableSettings.prototype, "selectAll", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getTableSettingPathV2(pageKeys)
        }
    })
], CommonTableSettings.prototype, "selectionLimit", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    }),
    decorators_1.validity({
        since: '1.75.2'
    })
], CommonTableSettings.prototype, "dataLoadSettings", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    }),
    decorators_1.validity({
        since: '1.42.0'
    })
], CommonTableSettings.prototype, "condensedTableLayout", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    }),
    decorators_1.validity({
        since: '1.56.0'
    })
], CommonTableSettings.prototype, "enableTableFilterInPageVariant", void 0);
exports.CommonTableSettings = CommonTableSettings;
class ResponsiveTable extends CommonTableSettings {
}
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getTableSettingPathV2(pageKeys)
        }
    })
], ResponsiveTable.prototype, "type", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    }),
    decorators_1.validity({
        since: '1.54.0'
    })
], ResponsiveTable.prototype, "createWithFilters", void 0);
__decorate([
    decorators_1.exportRule({
        flex: {
            controlId: getSmartTableControlId,
            controlType: () => types_1.ControlType.Table,
            exportFunction: decoration_1.exportToFlexChange
        }
    })
], ResponsiveTable.prototype, "popinLayout", void 0);
__decorate([
    decorators_1.exportRule({
        flex: {
            controlId: getSmartTableControlId,
            controlType: () => types_1.ControlType.Table,
            exportFunction: decoration_1.exportToFlexChange
        }
    })
], ResponsiveTable.prototype, "includeItemInSelection", void 0);
exports.ResponsiveTable = ResponsiveTable;
class TreeTable extends CommonTableSettings {
}
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getTableSettingPathV2(pageKeys)
        }
    })
], TreeTable.prototype, "type", void 0);
exports.TreeTable = TreeTable;
class AnalyticalTable extends CommonTableSettings {
}
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getTableSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "type", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "qualifier", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "smartVariantManagement", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "defaultContentView", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "defaultFilterMode", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "showGoButtonOnFilterBar", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "contentTitle", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "autoHide", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "filterDefaultsFromSelectionVariant", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    })
], AnalyticalTable.prototype, "allFiltersAsInParameters", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    }),
    decorators_1.validity({
        since: '1.54.0'
    })
], AnalyticalTable.prototype, "createWithFilters", void 0);
exports.AnalyticalTable = AnalyticalTable;
class GridTable extends CommonTableSettings {
}
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getTableSettingPathV2(pageKeys)
        }
    })
], GridTable.prototype, "type", void 0);
__decorate([
    decorators_1.exportRule({
        manifest: {
            path: (pageKeys) => application_1.getAppSettingPathV2(pageKeys)
        }
    }),
    decorators_1.validity({
        since: '1.54.0'
    })
], GridTable.prototype, "createWithFilters", void 0);
exports.GridTable = GridTable;
